carma_cooperative_perception/src/multiple_object_tracker_component.cpp

```diff
@@ -452,6 +453,92 @@ static auto predict_track_states(std::vector<Track> tracks, units::time::second_
   return tracks;
 }
 
+struct SemanticDistance2dScore
+{
+  template <typename Track, typename Detection>
+  auto operator()(const Track & track, const Detection & detection) const -> std::optional<float>
+  {
+    if constexpr (std::is_same_v<decltype(track.state), decltype(detection.state)>) {
+      const auto dist{two_dimensional_distance(track.state, detection.state)};
+      if (
+        track.semantic_class == mot::SemanticClass::kUnknown ||
+        detection.semantic_class == mot::SemanticClass::kUnknown) {
+        return dist;
+      }
+
+      if (track.semantic_class != detection.semantic_class) {
+        return std::nullopt;
+      }
+
+      return dist;
+    } else {
+      return std::nullopt;
+    }
```

Comment:
optional change, but we can reduce the if wrapping further down by returning nullopt if the first `if constexpr (std::is_same_v<decltype(track.state), decltype(detection.state)>)` doesn't hold.